src/dmd/expressionsem.d

```diff
@@ -2370,7 +2385,32 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
                 {
                     auto e = new AssignExp(gate.loc, new VarExp(gate.loc, gate), IntegerExp.createBool(true));
                     eprefix = Expression.combine(eprefix, e.expressionSemantic(sc));
-                    gate = null;
+                    if (!callerDestroysArgs)
+                        gate = null;
+                }
+            }
+            else if (callerDestroysArgs && i <= lastdtor)
```

Comment:
I think this should be doable without this. In new line 2352, we have the `edtor => (__gate || edtor)` rewrite. IIRC, the whole point of this `__gate` is to transfer ownership of all temporary args requiring destruction to the call**ee** just before the call (i.e., if all arguments were constructed successfully without throw). For `extern(D)`, it's up to the callee to destruct the parameter [which, low-level wise, is either the original argument passed by ref or a moved/blitted instance].

So if `callerDestroysArgs == true`, it should suffice to keep the original `edtor` and not declaring+setting `__gate`.